src/getopt.c

```diff
@@ -178,15 +189,27 @@ void opt_check(struct opt_entry *list, opt_flags flg, char **argv)
 				        opt_errors[res], *opt);
 			error();
 		}
-		/* Alter **argv to reflect to full option names */
+		/* Expand **argv to reflect the full option names */
 		else if (*opt[0] == '-') {
 			int len = strlen(completed) + 2 + 1;
+			if (completed_negated)
+				len += 3;
 			if (completed_param)
 				len += strlen(completed_param) + 1;
 			*opt = mem_alloc_tiny(len, MEM_ALIGN_NONE);
-			sprintf(*opt, "--%s%s%s", completed,
+			sprintf(*opt, "--%s%s%s%s",
+			        completed_negated ? "no-" : "",
+			        completed,
 			        completed_param ? "=" : "",
 			        completed_param ? completed_param : "");
 		}
 	}
+
+	/* Set all three-states */
+	if ((entry = list))
+	while (entry->name) {
+		if (entry->req_clr & OPT_THREESTATE)
+			*(int*)entry->param = entry->seen ? entry->threestate_bool : -1;
+		entry++;
```

Comment:
This first implementation requires `param` to be an integer. I guess we could implement string three-states if we wanted to, such as (NULL, empty string or non-empty string) mapping to (not given, given as `--no-opt`, given as `-opt=string`) but I'm not sure we'll need it so I'll fix that if/when we do.